src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCIBuildJobExecutionService.java

```diff
@@ -0,0 +1,621 @@
+     */
+    private final Path buildScriptFilePath;
+
+    /**
+     * Instead of creating a new XMLInputFactory for every build job, it is created once and provided as a Bean (see {@link LocalCIConfiguration#localCIXMLInputFactory()}).
+     */
+    private final XMLInputFactory localCIXMLInputFactory;
+
+    @Value("${artemis.version-control.url}")
+    private URL localVCBaseUrl;
+
+    @Value("${artemis.version-control.local-vcs-repo-path}")
+    private String localVCBasePath;
+
+    @Value("${artemis.continuous-integration.build.images.java.default}")
+    String dockerImage;
+
+    @Value("${artemis.continuous-integration.timeout-seconds:120}")
+    private int timeoutSeconds;
+
+    @Value("${artemis.continuous-integration.asynchronous:true}")
+    private boolean runBuildJobsAsynchronously;
+
+    public LocalCIBuildJobExecutionService(Optional<VersionControlService> versionControlService, ExecutorService localCIBuildExecutorService, DockerClient dockerClient,
+            ProgrammingMessagingService programmingMessagingService, TemplateProgrammingExerciseParticipationRepository templateProgrammingExerciseParticipationRepository,
+            SolutionProgrammingExerciseParticipationRepository solutionProgrammingExerciseParticipationRepository,
+            ProgrammingExerciseStudentParticipationRepository programmingExerciseStudentParticipationRepository, Path buildScriptFilePath, XMLInputFactory localCIXMLInputFactory) {
+        this.versionControlService = versionControlService;
+        this.localCIBuildExecutorService = localCIBuildExecutorService;
+        this.dockerClient = dockerClient;
+        this.programmingMessagingService = programmingMessagingService;
+        this.templateProgrammingExerciseParticipationRepository = templateProgrammingExerciseParticipationRepository;
+        this.solutionProgrammingExerciseParticipationRepository = solutionProgrammingExerciseParticipationRepository;
+        this.programmingExerciseStudentParticipationRepository = programmingExerciseStudentParticipationRepository;
+        this.buildScriptFilePath = buildScriptFilePath;
+        this.localCIXMLInputFactory = localCIXMLInputFactory;
+    }
+
+    /**
+     * Prepare paths to the assignment and test repositories and the build script and then submit the build job to the executor service.
+     *
+     * @param participation The participation of the repository for which the build job should be executed.
+     * @return A future that will be completed with the build result.
+     * @throws LocalCIException If the build job could not be submitted to the executor service.
+     */
+    public CompletableFuture<LocalCIBuildResult> addBuildJobToQueue(ProgrammingExerciseParticipation participation) {
+
+        ProjectType projectType = participation.getProgrammingExercise().getProjectType();
+        if (projectType == null || !projectType.isGradle()) {
```

Comment:
Since this validation is performed here, does this mean that it is possible to create a programming exercise whose type is not Gradle if the local CI profile is active?

Reply:
Yes, it should not be possible to create a programming exercise that has not of type Gradle, but I wanted to have an additional check at least once before the build job is queued.